### PR TITLE
Inverse transformation equations

### PR DESCRIPTION
**Pull request by szymag**
This PR introduce inverse transformation equations into `vector` module.

### CODE REVIEW

**Comment by szymag** [comment]
ping @Upabjojr

File: sympy/vector/coordsysrect.py [hunk 1]
```diff
@@ -162,6 +162,8 @@ def __new__(cls, name, location=None, rotation_matrix=None,
 
         obj._transformation_eqs = obj._x, obj._y, obj._y
 
+        obj._inv_transformation_eqs = obj._x, obj._y, obj._z
```
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: Shouldn't we use the `x`, `y`, `z` of the other coordinate system?

Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: I would suggest to use a `lambda`:
```
obj._inv_transformation_eqs = lambda x, y, z: (x, y, z)
```

**Comment by Upabjojr** [comment]
Looks good, except for the usage of _self.x, self.y, self.z_, it should rather be _sc.x, sc.y, sc.z_. You can either solve it through lambdas, or by creating a standard instance of `CoordSys3D` to be used as reference.

File: sympy/vector/coordsysrect.py [hunk 2]
```diff
@@ -196,21 +198,32 @@ def _connect_to_standard_cartesian(self, curv_coord_type):
...
-            self._transformation_eqs = curv_coord_type
-            self._h1, self._h2, self._h3 = self._calculate_lame_coefficients(curv_coord_type)
+            if any(i.atoms(BaseScalar) for i in curv_coord_type):
+                self._transformation_eqs = curv_coord_type
+                self._h1, self._h2, self._h3 = self._calculate_lame_coefficients(curv_coord_type)
+                if inverse:
+                    self._inv_transformation_eqs = self._calculate_inv_transformation_equations(
+                                                    self._transformation_equations())
+            else:
+                raise ValueError("Wrong set of parameter.")
```
Comment (szymag) [inline sympy/vector/coordsysrect.py]: I'm really dissatisfied from this condition, but at that moment don't have idea how to resolve that problem.

File: sympy/vector/coordsysrect.py [hunk 3]
```diff
@@ -196,21 +198,32 @@ def _connect_to_standard_cartesian(self, curv_coord_type):
...
         if isinstance(curv_coord_type, string_types):
             self._set_transformation_equations_mapping(curv_coord_type)
             self._set_lame_coefficient_mapping(curv_coord_type)
+            if inverse:
+                self._set_inv_trans_equations(curv_coord_type)
 
         elif isinstance(curv_coord_type, (tuple, list, Tuple)) and len(curv_coord_type) == 3:
-            self._transformation_eqs = curv_coord_type
-            self._h1, self._h2, self._h3 = self._calculate_lame_coefficients(curv_coord_type)
+            if any(i.atoms(BaseScalar) for i in curv_coord_type):
```
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: why this condition?

Comment (szymag) [inline sympy/vector/coordsysrect.py]: Because without that it could be possible to set transformation equations which doesn't depend on base scalars. 
For example, previously we could set transformation equations 
```python
a = CoordSys3D('a')
a._connect_to_standard_cartesian((x, y, z))
```
It shouldn't be possible. If someone wants to use symbols they should do
```python
a._connect_to_standard_cartesian(((x, y, z), (x, y, z)))
```

Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: Well, but if you do it that way, it can still be a problem. I mean, if only one of the three terms contains a base scalar, the transformation would basically contract two dimensions. Maybe it's better to open an issue and leave this for later. I think this could be connected with the orthogonality check (which is an open issue).

**Comment by szymag** [comment]
OK, maybe the solution will come itself

File: sympy/vector/coordsysrect.py (hunk 2 above)
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: We can re-address this in the future.

File: sympy/vector/coordsysrect.py [hunk 4]
```diff
@@ -240,6 +254,33 @@ def _set_transformation_equations_mapping(self, curv_coord_name):
...
+        curv_coord_name : str
+            The type of the new coordinate system.
+
+        """
+
+        equations_mapping = {
+            'cartesian': (self.x, self.y, self.z),
+            'spherical': (sqrt(self.x**2 + self.y**2 +self.z**2),
+                          acos((self.z) / sqrt(self.x**2 + self.y**2 + self.z**2)),
+                          atan(self.y / self.x)),
```
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: Can you use `atan2` instead of `atan`?

Comment (szymag) [inline sympy/vector/coordsysrect.py]: Sure, no problem.

**Comment by szymag** [comment]
@Upabjojr, I don't know why travis fail. I've run tests locally on my computer many times and everything was OK.
If you don't mind, I'd like to close this PR and open new one.

**Comment by Upabjojr** [comment]
@szymag you didn't update all the `atan` functions.

**Comment by Upabjojr** [comment]
`atan2` has the advantage that it determines the correct angle (up to 360 degrees), because `y` and `x` are separate arguments, so you have 4 sign combinations (x and y can be positive or negative separately).

The function `atan(y / x)` is a bit ambiguous, it determines the angle up to 180 degrees. `y / x` has only two possible signs (positive / negative).

**Comment by Upabjojr** [comment]
```python
    342     assert a._transformation_equations() == (a.x, a.y, a.z)
    343     assert a.lame_coefficients() == (1, 1, 1)
--> 344     assert a._inverse_transformation_equations() == (a.x, a.y, a.z)
    345     a._connect_to_standard_cartesian(((x, y, z), ((x * cos(y), x * sin(y), z))), inverse=False)
    346     assert a._transformation_equations() == (a.x * cos(a.y), a.x * sin(a.y), a.z)

ipdb> !a._inverse_transformation_equations() 
(a.z, a.y, a.x)
ipdb> !a._inverse_transformation_equations() == (a.x, a.y, a.z)
False
ipdb> !a._inverse_transformation_equations(), (a.x, a.y, a.z)
((a.z, a.y, a.x), (a.x, a.y, a.z))
```

**Comment by Upabjojr** [comment]
@szymag somehow the order of the base functions gets inverted.

**Comment by szymag** [comment]
There is still problem with the same line, `343` in `test_coordsysrect.py`.
I will try to open new PR.

**Comment by szymag** [comment]
I add `print` to see what is returning in this line. I hope it will works.

**Comment by szymag** [comment]
`Print` shows that the output is 
```python
(a.z, a.y, a.x)
```
instead of 
```python
(a.x, a.y, a.z)
```

**Comment by Upabjojr** [comment]
Maybe it's the output of `solve`? Maybe there is no definite order in its output. There should be an option to get the output as a dictionary.

**Comment by Upabjojr** [comment]
I think it's useless to open a new PR, the problem is likely in the code.

**Comment by szymag** [comment]
> Maybe it's the output of solve? Maybe there is no definite order in its output. There should be an option to get the output as a dictionary.

I will try it.
The strangest thing is that on my computer it works correctly and I have the same every file as here.

File: sympy/vector/coordsysrect.py [hunk 5]
```diff
@@ -431,7 +435,7 @@ def _transformation_equations(self):
         return self._transformation_eqs[:]
 
     def _inverse_transformation_equations(self):
-        return self._inv_transformation_eqs[:]
+        return self._inv_transformation_eqs
```
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: Why this?

**Comment by Upabjojr** [comment]
On my computer, I get the error. This can happen sometimes, one common cause is when you have an array that is read out of a dictionary or a set, without sorting. Python does not have a unique way of hashing objects (for example, Python functions are hashed by their memory address, which can vary every time you run the script), so the order may depend on the computer/operating system.

File: sympy/vector/coordsysrect.py (hunk 5 above)
Comment (szymag) [inline sympy/vector/coordsysrect.py]: Sorry, I revert this change.

**Comment by Upabjojr** [comment]
(I mean... elements in dict/set are stored in a Python internal data structure, their hash contributes to the way they are stored, so if you transform a dict into a list or array, the order may depend on the hash).

**Comment by szymag** [comment]
> On my computer, I get the error. 

It's better that this error occurs now.
